src/Controller/ControllerFactory.php

```diff
@@ -0,0 +1,149 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.3.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Controller;
+
+use Cake\Core\App;
+use Cake\Http\ControllerFactoryInterface;
+use Cake\Http\Exception\MissingControllerException;
+use Cake\Http\ServerRequest;
+use Cake\Utility\Inflector;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use ReflectionClass;
+
+/**
+ * Factory method for building controllers for request.
+ *
+ * @implements \Cake\Http\ControllerFactoryInterface<\Cake\Controller\Controller>
+ */
+class ControllerFactory implements ControllerFactoryInterface
+{
+    /**
+     * Create a controller for a given request.
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request to build a controller for.
+     * @return \Cake\Controller\Controller
+     * @throws \Cake\Http\Exception\MissingControllerException
+     * @throws \ReflectionException
+     */
+    public function create(ServerRequestInterface $request): Controller
+    {
+        $className = $this->getControllerClass($request);
+        if (!$className) {
+            $this->missingController($request);
+        }
+        /** @var string $className */
+        $reflection = new ReflectionClass($className);
+        if ($reflection->isAbstract() || $reflection->isInterface()) {
+            $this->missingController($request);
+        }
+
+        /** @var \Cake\Controller\Controller $controller */
+        $controller = $reflection->newInstance($request);
```

Comment:
This is a nice simplification. The controller can create its own response as needed.